lib/pbench/agent/tool_meister_client.py

```diff
@@ -41,195 +34,173 @@ class Client:
+        metadata = dict(group=group, directory=directory, args=args)
+        self.logger.debug(
+            "publish state signal for state %s with metadata: %s", action, metadata
+        )
         try:
-            num_present = self.redis_server.publish(
-                self.from_client_channel, json.dumps(msg)
+            resp, msgs = self.sig_pub.publish_signal(
+                event=action, tag="from_pbench_client", metadata=metadata, timeout=100
             )
         except Exception:
-            self.logger.exception("Failed to publish client message")
+            self.logger.exception("Failed to publish client signal")
             return 1
         else:
-            self.logger.debug("published %s", self.from_client_channel)
-            if num_present != 1:
-                self.logger.error(
-                    "Failed to publish to the TDS, encountered %d subscribers"
-                    " on the channel",
-                    num_present,
-                )
+            if resp != 0:
+                self.logger.error("Missing or bad response from the TDS, %s", str(resp))
                 ret_val = 1
+                for responder, msg in msgs.items():
+                    if msg != "success":
+                        self.logger.warning(
+                            "TDS responder %s reported: %s", responder, msg
+                        )
             else:
                 ret_val = 0
-
-        # Wait for an operational status message from the Tool Data Sink
-        # reporting the combined response of the Tool Meisters.
-        for data in self.to_client_chan.fetch_json(self.logger):
-            try:
-                kind = data["kind"]
-                action_r = data["action"]
-                status = data["status"]
-            except Exception:
-                self.logger.error("unrecognized status payload, %r", data)
-                ret_val = 1
-            else:
-                if kind != "ds":
-                    self.logger.warning("unrecognized kind in payload, %r", data)
-                    ret_val = 1
-                    continue
-                if action_r != action:
-                    self.logger.warning("unrecognized action in payload, %r", data)
-                    ret_val = 1
-                    continue
-                if status != "success":
-                    self.logger.warning(
-                        "Status message not successful: '%s'",
-                        status,
-                    )
-                    ret_val = 1
-                break
         return ret_val
 
-    def terminate(self, group, interrupt=False):
+    def publish(self, group: str, directory: str, action: str, args=None) -> int:
+        """publish a state signal formed from the group, directory,
+        action, and args arguments.
+
+        Returns 0 on success, 1 on failure; logs are also written for any
+        errors encountered.
+        """
+        if action not in tm_allowed_actions:
```

Comment:
We have moved the bulk of the code out of this method into a helper, but we are only doing the action test here?  Is it worth making this change?

Reply:
Yeah, I believe you had explained to me earlier in the year that it would be good practice to separate the mechanism of actually using the `SignalExporter` for publishing/response collecting into a helper, while leaving all additional functionality of the client publish in the original method. That way, if anything needs to change in the `SignalExporter`/actual publishing mechanism, the helper can be changed and the actual method remains consistent, and if we want to change Pbench Client specific checks/allowances/features, the original method can change and the mechanism helper can remain constant.

Reply:
Sure that makes sense, but ... just to avoid one test it doesn't seem worth it.

Reply:
Should I recombine them into one then?